### PR TITLE
add ComposeSorts and CompareRegionData helpers

### PR DESCRIPTION
**Pull request by emptyrivers**
# Description

Perhaps will make custom sort functions a bit easier to manage. To use, write something like this in custom sort:

```lua
WeakAuras.CompareRegionData({"region", "state", "expirationTime"})
```

Seriously, that's it. There isn't even any need to wrap it in a `function(a, b) ... end`, though of course you can do so. In which case I would recommend caching the function that you get from CompareRegionData into `aura_env`, and then calling it yourself on the last line.

If you want to compare more than one value in the regionData object, you can compose them:

```lua
WeakAuras.ComposeSorts(
  WeakAuras.CompareRegionData({"region", "state", "expirationTime"}),
  WeakAuras.CompareRegionData({"region", "state", "index"})
)
```

By default, `CompareRegionData` will use the `<` operator to compare values. You can override this by providing a function as the second parameter, which will be used instead.

ComposeSorts assumes that the functions it gets are passed in priority order, and a result of "equal" (or maybe "incomparable") is indicated by a function returning a nil value. If a compare function returns nil, then the next function is checked. And then the next, and so on down the list in the same order they were passed in.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

The Great and Wonderful Tester Extraordinaire @Causese said he couldn't find any issues, and he likes the feature.

# Checklist:

- [x] Finish converting builtin sort options to use helpers themselves

### CODE REVIEW

**Comment by InfusOnWoW** [comment]
Code wise I'd very much in favor of this clean up. Though we need to be figure out all the behaviour differences and whether they are acceptable.

File: WeakAuras/RegionTypes/DynamicGroup.lua [hunk 1]
```diff
@@ -115,72 +115,112 @@ function WeakAuras.GetPolarCoordinates(x, y, originX, originY)
...
   none = function(data)
-    return function(a, b)
-      if a.dataIndex == b.dataIndex then
-        local aIndex = a.region.state and a.region.state.index
-        local bIndex = b.region.state and b.region.state.index
-        if bIndex and aIndex then
-          if type(aIndex) ~= type(bIndex) then
-            -- state.index can be any value from custom code,
-            -- so guard against disparate types which can't be compared
-            return type(aIndex) < type(bIndex)
```
[the remark this answers is not in the document]
Comment (emptyrivers) [reply WeakAuras/RegionTypes/DynamicGroup.lua]: I re-implemented this check, for sort stability's sake.

**Comment by emptyrivers** [comment]
I did intentionally add one behavior change here. On ascending sorts, an `expirationTime` of `nil` now sorts before every non nil `expirationTime`, and vice versa for descending.

**Comment by emptyrivers** [comment]
We've gotten feedback from users that having nil expirationTime be last on ascending is confusing, and I've always sort of agreed. Since I was in the area, I went ahead and changed it. If users disagree with this, then it is very simple to tune the behavior, by using `WeakAuras.SortNilLast` instead of `WeakAuras.SortNilFirst`.

File: WeakAuras/RegionTypes/DynamicGroup.lua [hunk 2]
```diff
@@ -115,72 +115,112 @@ function WeakAuras.GetPolarCoordinates(x, y, originX, originY)
...
-          elseif aIndex == bIndex then
-            return nil
-          else
-            return aIndex < bIndex
-          end
-        elseif aIndex then
```
[the remark this answers is not in the document]
Comment (Stanzilla) [reply WeakAuras/RegionTypes/DynamicGroup.lua]: @emptyrivers are we keeping this difference then?

[the remark this answers is not in the document]
Comment (emptyrivers) [reply WeakAuras/RegionTypes/DynamicGroup.lua]: I'm not sure. To be honest I didn't really follow infus' comment, and I'm not sure if the rework currently maintains this behavior or not.

**Pull request merged**